Android Builds Failing in Latest GH Actions

### Describe the bug

https://github.com/microsoft/msquic/actions/runs/13514388510/job/37760398206?pr=4853

```
In file included from <built-in>:398:
<command line>:23:9: warning: '__ANDROID_API__' macro redefined [-Wmacro-redefined]
   23 | #define __ANDROID_API__ 29
      |         ^
<built-in>:390:9: note: previous definition is here
  390 | #define __ANDROID_API__ __ANDROID_MIN_SDK_VERSION__
      |         ^
In file included from <built-in>:398:
<command line>:23:9: warning: '__ANDROID_API__' macro redefined [-Wmacro-redefined]
   23 | #define __ANDROID_API__ 29
      |         ^
...
```

From past experience, this is caused by GitHub build machines changing Android versions, but it may be a different issue this time.

### Affected OS

- [ ] Windows
- [x] Linux
- [ ] macOS
- [ ] Other (specify below)

### Additional OS information

_No response_

### MsQuic version

main

### Steps taken to reproduce bug

Create a PR

### Expected behavior

All CI builds pass

### Actual outcome

Android is failing.

### Additional details

_No response_

There were few errors related to unused-result:

/home/runner/work/msquic/msquic/src/tools/sample/sample.c:667:5: error: ignoring return value of function declared with 'warn_unused_result' attribute [-Werror,-Wunused-result]
  667 |     getchar();
      |     ^~~~~~~
1 error generated.
gmake[2]: *** [src/tools/sample/CMakeFiles/quicsample.dir/build.make:79: src/tools/sample/CMakeFiles/quicsample.dir/sample.c.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:1171: src/tools/sample/CMakeFiles/quicsample.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....
gmake[2]: Leaving directory '/home/runner/work/msquic/msquic/build/android/x64_openssl'
/home/runner/work/msquic/msquic/src/tools/interopserver/InteropServer.cpp:137:13: error: ignoring return value of function declared with 'warn_unused_result' attribute [-Werror,-Wunused-result]
  137 |             getchar();
      |             ^~~~~~~
1 error generated.


These are fixed in draft PR, but unable to proceed to build due to tool chain issue with OpenSSL3 on Android (NDK) as below:

Configuring OpenSSL version 3.1.4+quic for target android-x86_64
Using os-specific seed configuration
Failure!  build file wasn't produced.
gmake[2]: Leaving directory '/home/runner/work/msquic/msquic/build/android/x64_openssl3'
Please read INSTALL.md and associated NOTES-* files.  You may also have to
look over your available compiler tool chain or change your configuration.

**no NDK x86_64-linux-android-gcc on $PATH at (eval 10) line 142.**
gmake[2]: *** [_deps/opensslquic-build/CMakeFiles/OpenSSL_Target.dir/build.make:81: _deps/opensslquic-build/submodules/openssl3/Makefile] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:700: _deps/opensslquic-build/CMakeFiles/OpenSSL_Target.dir/all] Error 2
gmake[1]: *** Waiting for unfinished jobs....



When OpenSSL 3.1 was being configured, it started looking for x86_64-linux-android-gcc, I think it should start using Clang instead of gcc; probably this is some configuration issue on OpenSSL3.1 on Android.

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -120,9 +120,6 @@ jobs:
         # Android doesn't use system crypto
         - plat: android
           systemcrypto: '-UseSystemOpenSSLCrypto'
-        # TODO: Fix android on ubuntu-24.04
-        - plat: android
-          os: 'ubuntu-24.04'
         # No openssl system crypto on ubuntu-22.04
         - plat: linux
           os: 'ubuntu-22.04'

--- a/scripts/build.ps1
+++ b/scripts/build.ps1
@@ -507,16 +507,22 @@ function CMake-Generate {
         $Arguments += " -DQUIC_HIGH_RES_TIMERS=on"
     }
     if ($Platform -eq "android") {
-        $NDK = $env:ANDROID_NDK_LATEST_HOME -replace '26\.\d+\.\d+', '25.2.9519653' # Temporary work around. Use RegEx to replace newer version.
+        $NDK = $env:ANDROID_NDK_LATEST_HOME -replace '26\.\d+\.\d+', '28.0.13004108' # Temporary work around. Use RegEx to replace newer version.
         $env:PATH = "$NDK/toolchains/llvm/prebuilt/linux-x86_64/bin:$env:PATH"
+        $env:ANDROID_NDK_HOME = $NDK
+        $env:ANDROID_NDK_ROOT = $NDK  # Set ANDROID_NDK_ROOT to ensure consistency
+        Write-Host "Set ANDROID_NDK_HOME: $env:ANDROID_NDK_HOME"
+        Write-Host "Set ANDROID_NDK_ROOT: $env:ANDROID_NDK_ROOT"
+
         switch ($Arch) {
             "x86"   { $Arguments += " -DANDROID_ABI=x86"}
             "x64"   { $Arguments += " -DANDROID_ABI=x86_64" }
             "arm"   { $Arguments += " -DANDROID_ABI=armeabi-v7a" }
             "arm64" { $Arguments += " -DANDROID_ABI=arm64-v8a" }
         }
+        $env:CC = "x86_64-linux-android29-clang"
+        $env:CXX = "x86_64-linux-android29-clang++"
         $Arguments += " -DANDROID_PLATFORM=android-29"
-        $env:ANDROID_NDK_HOME = $NDK
         $NdkToolchainFile = "$NDK/build/cmake/android.toolchain.cmake"
         $Arguments += " -DANDROID_NDK=""$NDK"""
         $Arguments += " -DCMAKE_TOOLCHAIN_FILE=""$NdkToolchainFile"""

--- a/scripts/build.ps1
+++ b/scripts/build.ps1
@@ -507,12 +507,14 @@ function CMake-Generate {
         $Arguments += " -DQUIC_HIGH_RES_TIMERS=on"
     }
     if ($Platform -eq "android") {
-        $NDK = $env:ANDROID_NDK_LATEST_HOME -replace '26\.\d+\.\d+', '28.0.13004108' # Temporary work around. Use RegEx to replace newer version.
+        #$NDK = $env:ANDROID_NDK_LATEST_HOME -replace '26\.\d+\.\d+', '28.0.13004108' # Temporary work around. Use RegEx to replace newer version.
+        $NDK = $env:ANDROID_NDK_LATEST_HOME
         $env:PATH = "$NDK/toolchains/llvm/prebuilt/linux-x86_64/bin:$env:PATH"
         $env:ANDROID_NDK_HOME = $NDK
         $env:ANDROID_NDK_ROOT = $NDK  # Set ANDROID_NDK_ROOT to ensure consistency
         Write-Host "Set ANDROID_NDK_HOME: $env:ANDROID_NDK_HOME"
         Write-Host "Set ANDROID_NDK_ROOT: $env:ANDROID_NDK_ROOT"
+        Write-Host "Set ANDROID_NDK_LATEST_HOME: $env:ANDROID_NDK_LATEST_HOME"
 
         switch ($Arch) {
             "x86"   { $Arguments += " -DANDROID_ABI=x86"}

--- a/scripts/build.ps1
+++ b/scripts/build.ps1
@@ -520,8 +520,6 @@ function CMake-Generate {
             "arm"   { $Arguments += " -DANDROID_ABI=armeabi-v7a" }
             "arm64" { $Arguments += " -DANDROID_ABI=arm64-v8a" }
         }
-        $env:CC = "x86_64-linux-android29-clang"
-        $env:CXX = "x86_64-linux-android29-clang++"
         $Arguments += " -DANDROID_PLATFORM=android-29"
         $NdkToolchainFile = "$NDK/build/cmake/android.toolchain.cmake"
         $Arguments += " -DANDROID_NDK=""$NDK"""

--- a/scripts/build.ps1
+++ b/scripts/build.ps1
@@ -507,7 +507,6 @@ function CMake-Generate {
         $Arguments += " -DQUIC_HIGH_RES_TIMERS=on"
     }
     if ($Platform -eq "android") {
-        #$NDK = $env:ANDROID_NDK_LATEST_HOME -replace '26\.\d+\.\d+', '28.0.13004108' # Temporary work around. Use RegEx to replace newer version.
         $NDK = $env:ANDROID_NDK_LATEST_HOME
         $env:PATH = "$NDK/toolchains/llvm/prebuilt/linux-x86_64/bin:$env:PATH"
         $env:ANDROID_NDK_HOME = $NDK

--- a/src/tools/forwarder/forwarder.cpp
+++ b/src/tools/forwarder/forwarder.cpp
@@ -234,6 +234,6 @@ int QUIC_MAIN_EXPORT main(int argc, char **argv) {
     CXPLAT_FRE_ASSERT(QUIC_SUCCEEDED(Listener.Start(Alpn, QuicAddr(QUIC_ADDRESS_FAMILY_UNSPEC, FrontEndPort))));
 
     printf("Press Enter to exit.\n\n");
-    getchar();
+    (void)getchar();
     return 0;
 }

--- a/src/tools/interopserver/InteropServer.cpp
+++ b/src/tools/interopserver/InteropServer.cpp
@@ -134,7 +134,7 @@ main(
         HttpServer Server(Registration, SupportedALPNs, ARRAYSIZE(SupportedALPNs), &ListenAddr, SslKeyLogFileParam);
         if (!GetFlag(argc, argv, "noexit")) {
             printf("Press Enter to exit.\n\n");
-            getchar();
+            (void)getchar();
         } else {
             CXPLAT_EVENT Event;
             CxPlatEventInitialize(&Event, TRUE, FALSE);

--- a/src/tools/ip/server/quicipserver.cpp
+++ b/src/tools/ip/server/quicipserver.cpp
@@ -182,7 +182,7 @@ RunServer(
     }
 
     printf("Press Enter to exit.\n\n");
-    getchar();
+    (void)getchar();
 
 Error:
 

--- a/src/tools/lb/loadbalancer.cpp
+++ b/src/tools/lb/loadbalancer.cpp
@@ -220,7 +220,7 @@ main(int argc, char **argv)
     PublicInterface = new LbPublicInterface(&PublicAddr);
 
     printf("Press Enter to exit.\n\n");
-    getchar();
+    (void)getchar();
 
     delete PublicInterface;
     CxPlatDataPathUninitialize(Datapath);

--- a/src/tools/sample/sample.c
+++ b/src/tools/sample/sample.c
@@ -664,7 +664,7 @@ RunServer(
     // Continue listening for connections until the Enter key is pressed.
     //
     printf("Press Enter to exit.\n\n");
-    getchar();
+    (void)getchar();
 
 Error:
 

